go/decoder.go

```diff
@@ -0,0 +1,597 @@
+package cbor
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"math"
+	"math/big"
+	"reflect"
+	"strings"
+)
+
+type TagDecoder interface {
+	// Handle things which match this.
+	//
+	// Setup like this:
+	// var dec Decoder
+	// var myTagDec TagDecoder
+	// dec.TagDecoders[myTagDec.GetTag()] = myTagDec
+	GetTag() uint64
+
+	// Sub-object will be decoded onto the returned object.
+	DecodeTarget() interface{}
+
+	// Run after decode onto DecodeTarget has happened.
+	// The return value from this is returned in place of the
+	// raw decoded object.
+	PostDecode(interface{}) (interface{}, error)
+}
+
+type Decoder struct {
+	rin io.Reader
+
+	// tag byte
+	c []byte
+
+	// many values fit within the next 8 bytes
+	b8 []byte
+
+	// Extra processing for CBOR TAG objects.
+	TagDecoders map[uint64]TagDecoder
+}
+
+func NewDecoder(r io.Reader) *Decoder {
+	return &Decoder{
+		rin:         r,
+		c:           make([]byte, 1),
+		b8:          make([]byte, 8),
+		TagDecoders: make(map[uint64]TagDecoder),
+	}
+}
+func (dec *Decoder) Decode(v interface{}) error {
+	rv := reflect.ValueOf(v)
+
+	return dec.DecodeAny(newReflectValue(rv))
+}
+
+func (dec *Decoder) DecodeAny(v DecodeValue) error {
+	var didread int
+	var err error
+
+	_, err = io.ReadFull(dec.rin, dec.c)
+
+	if didread == 1 {
+		/* log.Printf("got one %x\n", dec.c[0]) */
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if err := v.Prepare(); err != nil {
+		return err
+	}
+
+	return dec.innerDecodeC(v, dec.c[0])
+}
+
+func (dec *Decoder) handleInfoBits(cborInfo byte) (uint64, error) {
+	var aux uint64
+
+	if cborInfo <= 23 {
+		aux = uint64(cborInfo)
+		return aux, nil
+	} else if cborInfo == int8Follows {
+		didread, err := io.ReadFull(dec.rin, dec.b8[:1])
+		if didread == 1 {
+			aux = uint64(dec.b8[0])
+		}
+		return aux, err
+	} else if cborInfo == int16Follows {
+		didread, err := io.ReadFull(dec.rin, dec.b8[:2])
+		if didread == 2 {
```

Comment:
this sort of error handling seems weird to me... why arent we checking the error? `ReadFull` is guaranteed to read as much as we ask it to, and will error in any other case.